From: Andrew Morton <akpm@linux-foundation.org>
Date: Tue, 11 Aug 2026 17:23:28 -0700
Subject: [PATCH] mm-hmmc-hmm_do_fault-suppress-sparse-warning-fix

WARN_ON_ONCE() if the handler didn't set ->locked, per Stanislav

Link: https://lore.kernel.org/anu1N-DOnQwxO1kF@skinsburskii
Cc: David Hildenbrand <david@kernel.org>
Cc: kernel test robot <lkp@intel.com>
Cc: Stanislav Kinsburskii <skinsburskii@gmail.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---
 mm/hmm.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/mm/hmm.c b/mm/hmm.c
index 6ce1b409ece7d..2f1e98c6b6440 100644
--- a/mm/hmm.c
+++ b/mm/hmm.c
@@ -670,8 +670,10 @@ static int hmm_do_fault(struct mm_struct *mm,
 		ret = handle_mm_fault(vma, addr, fault_flags, NULL);
 
 		if (ret & (VM_FAULT_COMPLETED | VM_FAULT_RETRY)) {
-			if (hmm_vma_walk->locked)	/* needed by sparse */
+			if (hmm_vma_walk->locked)    /* needed by sparse */
 				*hmm_vma_walk->locked = false;
+			else
+				WARN_ON_ONCE(1);    /* broken fault handler */
 			return HMM_FAULT_UNLOCKED;
 		}
 
-- 
2.47.3

